lib/assertions.js

```diff
@@ -1237,7 +1246,10 @@ module.exports = function (expect) {
         keys.forEach(function (key, index) {
             promiseByKey[key] = expect.promise(function () {
                 var valueKeyType = expect.findTypeOf(value[key]);
-                if (valueKeyType.is('function')) {
+                if (valueKeyType.is('expect.it')) {
+                    expect.context.thisObject = subject;
```

Comment:
Yeah, I don't think we can use a single field to keep track of the `thisObject` at all levels:

```js
var expect = require('./');

expect.addAssertion('<any> when delayed <assertion?>', function (expect) {
    return expect.promise.delay(10).then(() => expect.shift());
});

function Greeter(name) {
    this.name = name;
}

Greeter.prototype.greet = function (greeting) {
    return greeting + ', ' + this.name;
};

Greeter.prototype.getFareweller = function () {
    return {
        prefix: 'goodbye',
        name: '>>' + this.name + '<<',
        bye: function () {
            return this.prefix + ', ' + this.name;
        }
    };
};

it('should foo', function () {
    var johnDoeGreeter = new Greeter('John Doe');

    return expect(johnDoeGreeter, 'to satisfy', {
        getFareweller: expect.it('when called when delayed to satisfy', {
            bye: expect.it('when called to equal', 'goodbye, >>John Doe<<')
        }),
        greet: expect.it('when delayed when called with', ['hello'], 'to equal', 'hello, John Doe')
    });
});
```

Yields:

```
  Error
      at Function.callInNestedContext (lib/createWrappedExpectProto.js:120:40)
      at Function.fail (lib/createWrappedExpectProto.js:74:18)
      at lib/assertions.js:1279:24
      at tryCatcher (node_modules/unexpected-bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (node_modules/unexpected-bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (node_modules/unexpected-bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (node_modules/unexpected-bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (node_modules/unexpected-bluebird/js/main/async.js:123:16)
      at Async._drainQueues (node_modules/unexpected-bluebird/js/main/async.js:133:10)
      at Async.drainQueues (node_modules/unexpected-bluebird/js/main/async.js:15:14)
      at lib/workQueue.js:7:13
      at Array.forEach (native)
      at Object.drain (lib/workQueue.js:6:20)
      at Immediate.<anonymous> (lib/workQueue.js:16:19)
```